File: claims_audit/rules.py
"""Rule-set loading and typed rule models.

Rules are authored as machine-readable YAML in ``rules/audit_rules.yaml`` and
loaded into typed ``Rule`` objects. The YAML is the single source of truth: the
deterministic engine, both agents' ``lookup_rule`` tool, and the eval all read
the same file. This mirrors how a statute/compliance engine keeps its logic in
data rather than code.
"""

from __future__ import annotations

from enum import Enum
from pathlib import Path

import yaml
from pydantic import BaseModel, Field

from claims_audit.models import DefectType, Severity

DEFAULT_RULES_PATH = Path(__file__).resolve().parent.parent / "rules" / "audit_rules.yaml"


class RuleType(str, Enum):
    """Kinds of edits the engine knows how to evaluate."""

    NCCI_PAIR = "ncci_pair"
    MUE_UNITS = "mue_units"
    DUPLICATE = "duplicate"
    MODIFIER_REQUIRED = "modifier_required"
    UPCODING_EM = "upcoding_em"


class Rule(BaseModel):
    """One machine-readable audit rule."""

    id: str
    type: RuleType
    defect_type: DefectType
    severity: Severity = Severity.MEDIUM
    description: str
    # Free-form typed params; each RuleType interprets its own keys.
    params: dict = Field(default_factory=dict)


class RuleSet(BaseModel):
    """A collection of rules with id-based lookup."""

    rules: list[Rule]

    def by_id(self, rule_id: str) -> Rule | None:
        for r in self.rules:
            if r.id == rule_id:
                return r
        return None

    def ids(self) -> set[str]:
        return {r.id for r in self.rules}

    def of_type(self, rule_type: RuleType) -> list[Rule]:
        return [r for r in self.rules if r.type == rule_type]

    def __iter__(self):
        return iter(self.rules)

    def __len__(self) -> int:
        return len(self.rules)


def load_rules(path: str | Path | None = None) -> RuleSet:
    """Load and validate the rule set from YAML."""
    p = Path(path) if path else DEFAULT_RULES_PATH
    raw = yaml.safe_load(p.read_text(encoding="utf-8"))
    rules = [Rule(**item) for item in raw["rules"]]
    ids = [r.id for r in rules]
    if len(ids) != len(set(ids)):
        dupes = sorted({i for i in ids if ids.count(i) > 1})
        raise ValueError(f"Duplicate rule ids in {p}: {dupes}")
    return RuleSet(rules=rules)

File: agent_aisdk/src/rules.ts
/**
 * Rule-set loading — mirrors `claims_audit/rules.py`. The YAML in
 * `rules/audit_rules.yaml` is the single source of truth; the deterministic
 * engine, both Python agents, and this arm all read the same file. We parse it
 * (never re-encode it) so there is no chance of rule drift between substrates.
 */
import { readFileSync } from 'node:fs';
import { parse } from 'yaml';
import { z } from 'zod';
import { DefectTypeSchema, SeveritySchema } from './models.js';
import { RULES_PATH } from './paths.js';

export const RuleTypeSchema = z.enum([
  'ncci_pair',
  'mue_units',
  'duplicate',
  'modifier_required',
  'upcoding_em',
]);
export type RuleType = z.infer<typeof RuleTypeSchema>;

export const RuleSchema = z
  .object({
    id: z.string(),
    type: RuleTypeSchema,
    defect_type: DefectTypeSchema,
    severity: SeveritySchema.default('medium'),
    description: z.string(),
    params: z.record(z.string(), z.unknown()).default({}),
  })
  .strict();
export type Rule = z.infer<typeof RuleSchema>;

const RuleFileSchema = z.object({ rules: z.array(RuleSchema) });

export class RuleSet {
  readonly rules: readonly Rule[];
  private readonly byIdMap: Map<string, Rule>;

  constructor(rules: readonly Rule[]) {
    const ids = rules.map((r) => r.id);
    const unique = new Set(ids);
    if (unique.size !== ids.length) {
      const dupes = [...unique].filter((id) => ids.filter((x) => x === id).length > 1).sort();
      throw new Error(`Duplicate rule ids: ${dupes.join(', ')}`);
    }
    this.rules = rules;
    this.byIdMap = new Map(rules.map((r) => [r.id, r]));
  }

  byId(ruleId: string | undefined | null): Rule | undefined {
    if (ruleId == null) return undefined;
    return this.byIdMap.get(ruleId);
  }

  ids(): Set<string> {
    return new Set(this.byIdMap.keys());
  }

  [Symbol.iterator](): Iterator<Rule> {
    return this.rules[Symbol.iterator]();
  }

  get length(): number {
    return this.rules.length;
  }
}

export function loadRules(path: string = RULES_PATH): RuleSet {
  const raw: unknown = parse(readFileSync(path, 'utf-8'));
  const parsed = RuleFileSchema.parse(raw);
  return new RuleSet(parsed.rules);
}
